### PR TITLE
Adding password policy pattern feature

### PR DESCRIPTION
**Pull request by GayanM**

### CODE REVIEW

File: components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/internal/PasswordPolicyPatternServiceComponent.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+        if (log.isDebugEnabled()) {
+            log.debug("Policy enforcer service registered successfully.");
+        }
+    }
+
+    protected void unregisterPolicyEnforcer(PolicyEnforcer service) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Data source service unregistered.");
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/internal/PasswordPolicyPatternServiceComponent.java]: This log doesn't make sense here.

File: components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                    " characters") {
+
+                @Override
+                public boolean passRule(char[] password) {
+                    return password.length <= new PasswordPolicyBean().getMaxLength();
+                }
+
+            },
+
+            new BaseRule("Password needs at least an upper case character") {
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java]: at least 'one' would have been better.. :)

File: components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    @Override
+    public ValidationResult enforce(Object... args) {
+
+        char[] password = (char[]) args[0];
+        StringBuilder errorMessage = new StringBuilder();
+        ValidationResult validationResult = new ValidationResult();
+
+        //Give first priority to regex pattern
+        PasswordPolicyBean passwordPolicyBean = new PasswordPolicyBean();
+        if (!passwordPolicyBean.getRegex().isEmpty()) {
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java]: Can getRegex return null?

File: components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            Pattern pattern = Pattern.compile(passwordPolicyBean.getRegex());
+            if (pattern.matcher(CharBuffer.wrap(password)).matches()) {
+                validationResult.setSuccess(true);
+                return validationResult;
+            }
+            validationResult.setSuccess(false);
+            validationResult.setMessage("Password your entered does not adhere to password policy");
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java]: Password *you* entered. --> Entered password does not adhere to password policy.

File: components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java (hunk 3 above)
Comment (GayanM) [inline components/org.wso2.carbon.identity.policy.password/src/main/java/org/wso2/carbon/identity/policy/password/pattern/registry/impl/DefaultPasswordPatternPolicy.java]: no

**Pull request merged**